File: fetchAssemblyData.js
const axios = require('axios');
const fs = require('fs').promises;

const APIs = [
  {
    name: 'ALLNAMEMBER',
    url: 'https://open.assembly.go.kr/portal/openapi/ALLNAMEMBER',
    filename: 'assembly_members_integrated.json'
  },
  {
    name: 'nprlapfmaufmqytet',
    url: 'https://open.assembly.go.kr/portal/openapi/nprlapfmaufmqytet',
    filename: 'assembly_members_history_daesu_{DAESU}.json',
    isDaesuIteration: true
  },
  {
    name: 'nwvrqwxyaytdsfvhu',
    url: 'https://open.assembly.go.kr/portal/openapi/nwvrqwxyaytdsfvhu',
    filename: 'assembly_members_profile.json'
  },
  {
    name: 'nzmimeepazxkubdpn',
    url: 'https://open.assembly.go.kr/portal/openapi/nzmimeepazxkubdpn',
    filename: 'assembly_bills_age_{AGE}.json',
    isAgeIteration: true
  }
];

const API_KEY = '84d2bb829a0d413b97da6e9d2809db9e';
const PAGE_SIZE = 1000;

async function fetchAPIData(api) {
  console.log(`Starting data collection for ${api.name}...`);
  
  let allData = [];
  let pIndex = api.startIndex || 1;
  let hasMoreData = true;

  while (hasMoreData) {
    try {
      const url = `${api.url}?KEY=${API_KEY}&Type=json&pIndex=${pIndex}&pSize=${PAGE_SIZE}${api.extraParams || ''}`;
      console.log(`Fetching ${api.name} - Page ${pIndex}...`);
      
      const response = await axios.get(url);
      const data = response.data;
      
      // Extract the actual data array from the response
      const dataKey = Object.keys(data)[0];
      const items = data[dataKey]?.[1]?.row || [];
      
      if (items.length === 0) {
        hasMoreData = false;
        console.log(`No more data for ${api.name} at page ${pIndex}`);
      } else {
        allData = allData.concat(items);
        console.log(`Collected ${items.length} items from ${api.name} page ${pIndex}`);
        pIndex++;
      }
      
      // Add delay to avoid overwhelming the server
      await new Promise(resolve => setTimeout(resolve, 100));
      
    } catch (error) {
      console.error(`Error fetching ${api.name} page ${pIndex}:`, error.message);
      hasMoreData = false;
    }
  }

  return {
    api: api.name,
    totalItems: allData.length,
    data: allData
  };
}

async function fetchDaesuData(api, daesu) {
  console.log(`Starting data collection for ${api.name} DAESU ${daesu}...`);
  
  let allData = [];
  let pIndex = 1;
  let hasMoreData = true;

  while (hasMoreData) {
    try {
      const url = `${api.url}?KEY=${API_KEY}&Type=json&pIndex=${pIndex}&pSize=${PAGE_SIZE}&DAESU=${daesu}`;
      console.log(`Fetching ${api.name} DAESU ${daesu} - Page ${pIndex}...`);
      
      const response = await axios.get(url);
      const data = response.data;
      
      // Extract the actual data array from the response
      const dataKey = Object.keys(data)[0];
      const items = data[dataKey]?.[1]?.row || [];
      
      if (items.length === 0) {
        hasMoreData = false;
        console.log(`No more data for ${api.name} DAESU ${daesu} at page ${pIndex}`);
      } else {
        allData = allData.concat(items);
        console.log(`Collected ${items.length} items from ${api.name} DAESU ${daesu} page ${pIndex}`);
        pIndex++;
      }
      
      // Add delay to avoid overwhelming the server
      await new Promise(resolve => setTimeout(resolve, 100));
      
    } catch (error) {
      console.error(`Error fetching ${api.name} DAESU ${daesu} page ${pIndex}:`, error.message);
      hasMoreData = false;
    }
  }

  return {
    api: api.name,
    daesu: daesu,
    totalItems: allData.length,
    data: allData
  };
}

async function fetchAgeData(api, age) {
  console.log(`Starting data collection for ${api.name} AGE ${age}...`);
  
  let allData = [];
  let pIndex = api.startIndex || 1;
  let hasMoreData = true;

  while (hasMoreData) {
    try {
      const url = `${api.url}?KEY=${API_KEY}&Type=json&pIndex=${pIndex}&pSize=${PAGE_SIZE}&AGE=${age}`;
      console.log(`Fetching ${api.name} AGE ${age} - Page ${pIndex}...`);
      
      const response = await axios.get(url);
      const data = response.data;
      
      // Extract the actual data array from the response
      const dataKey = Object.keys(data)[0];
      const items = data[dataKey]?.[1]?.row || [];
      
      if (items.length === 0) {
        hasMoreData = false;
        console.log(`No more data for ${api.name} AGE ${age} at page ${pIndex}`);
      } else {
        allData = allData.concat(items);
        console.log(`Collected ${items.length} items from ${api.name} AGE ${age} page ${pIndex}`);
        pIndex++;
      }
      
      // Add delay to avoid overwhelming the server
      await new Promise(resolve => setTimeout(resolve, 100));
      
    } catch (error) {
      console.error(`Error fetching ${api.name} AGE ${age} page ${pIndex}:`, error.message);
      hasMoreData = false;
    }
  }

  return {
    api: api.name,
    age: age,
    totalItems: allData.length,
    data: allData
  };
}

async function saveToFile(result, filename) {
  try {
    await fs.writeFile(filename, JSON.stringify(result, null, 2), 'utf8');
    console.log(`Saved ${result.totalItems} items to ${filename}`);
  } catch (error) {
    console.error(`Error saving to ${filename}:`, error.message);
  }
}

async function main() {
  console.log('Starting API data aggregation...');
  
  for (const api of APIs) {
    try {
      if (api.isDaesuIteration) {
        // Handle DAESU iteration for nprlapfmaufmqytet API
        for (let daesu = 10; daesu <= 22; daesu++) {
          const result = await fetchDaesuData(api, daesu);
          const filename = api.filename.replace('{DAESU}', daesu);
          await saveToFile(result, filename);
          console.log(`Completed ${api.name} DAESU ${daesu}: ${result.totalItems} total items\n`);
        }
      } else if (api.isAgeIteration) {
        // Handle AGE iteration for nzmimeepazxkubdpn API
        for (let age = 10; age <= 22; age++) {
          const result = await fetchAgeData(api, age);
          const filename = api.filename.replace('{AGE}', age);
          await saveToFile(result, filename);
          console.log(`Completed ${api.name} AGE ${age}: ${result.totalItems} total items\n`);
        }
      } else {
        const result = await fetchAPIData(api);
        await saveToFile(result, api.filename);
        console.log(`Completed ${api.name}: ${result.totalItems} total items\n`);
      }
    } catch (error) {
      console.error(`Failed to process ${api.name}:`, error.message);
    }
  }
  
  console.log('All APIs processed successfully!');
}

main().catch(console.error);

File: fetchAssemblyData.py
import requests
import json
import asyncio
import aiohttp
import time
import os
from pathlib import Path
from typing import List, Dict, Any, Optional
from dotenv import load_dotenv

load_dotenv()

class AssemblyDataFetcher:
    def __init__(self):
        self.api_key = os.getenv('API_KEY')
        if not self.api_key:
            raise ValueError('API_KEY is not set in environment variables. Please check your .env file.')
        self.page_size = 1000
        self.base_dir = Path(__file__).parent
        
        self.apis = [
            {
                'name': 'ALLNAMEMBER',
                'url': 'https://open.assembly.go.kr/portal/openapi/ALLNAMEMBER',
                'filename': 'assembly_members_integrated.json'
            },
            {
                'name': 'nprlapfmaufmqytet',
                'url': 'https://open.assembly.go.kr/portal/openapi/nprlapfmaufmqytet',
                'filename': 'assembly_members_history_daesu_{DAESU}.json',
                'is_daesu_iteration': True
            },
            {
                'name': 'nwvrqwxyaytdsfvhu',
                'url': 'https://open.assembly.go.kr/portal/openapi/nwvrqwxyaytdsfvhu',
                'filename': 'assembly_members_profile.json'
            },
            {
                'name': 'nzmimeepazxkubdpn',
                'url': 'https://open.assembly.go.kr/portal/openapi/nzmimeepazxkubdpn',
                'filename': 'assembly_bills_age_{AGE}.json',
                'is_age_iteration': True
            }
        ]

    async def fetch_api_data(self, session: aiohttp.ClientSession, api: Dict[str, Any], 
                           extra_params: str = '') -> Dict[str, Any]:
        """Fetch data from a single API endpoint with pagination"""
        print(f"Starting data collection for {api['name']}...")
        
        all_data = []
        p_index = api.get('start_index', 1)
        has_more_data = True

        while has_more_data:
            try:
                url = f"{api['url']}?KEY={self.api_key}&Type=json&pIndex={p_index}&pSize={self.page_size}{extra_params}"
                print(f"Fetching {api['name']} - Page {p_index}...")
                
                async with session.get(url) as response:
                    response.raise_for_status()
                    data = await response.json()
                    
                    # Extract the actual data array from the response
                    data_key = next(iter(data.keys()))
                    items = data.get(data_key, [None, {}])[1].get('row', []) if len(data.get(data_key, [])) > 1 else []
                    
                    if not items:
                        has_more_data = False
                        print(f"No more data for {api['name']} at page {p_index}")
                    else:
                        all_data.extend(items)
                        print(f"Collected {len(items)} items from {api['name']} page {p_index}")
                        p_index += 1
                    
                    # Add delay to avoid overwhelming the server
                    await asyncio.sleep(0.1)
                    
            except Exception as error:
                print(f"Error fetching {api['name']} page {p_index}: {error}")
                has_more_data = False

        return {
            'api': api['name'],
            'total_items': len(all_data),
            'data': all_data
        }

    async def fetch_daesu_data(self, session: aiohttp.ClientSession, api: Dict[str, Any], 
                             daesu: int) -> Dict[str, Any]:
        """Fetch data for a specific DAESU (assembly term)"""
        print(f"Starting data collection for {api['name']} DAESU {daesu}...")
        
        all_data = []
        p_index = 1
        has_more_data = True

        while has_more_data:
            try:
                url = f"{api['url']}?KEY={self.api_key}&Type=json&pIndex={p_index}&pSize={self.page_size}&DAESU={daesu}"
                print(f"Fetching {api['name']} DAESU {daesu} - Page {p_index}...")
                
                async with session.get(url) as response:
                    response.raise_for_status()
                    data = await response.json()
                    
                    # Extract the actual data array from the response
                    data_key = next(iter(data.keys()))
                    items = data.get(data_key, [None, {}])[1].get('row', []) if len(data.get(data_key, [])) > 1 else []
                    
                    if not items:
                        has_more_data = False
                        print(f"No more data for {api['name']} DAESU {daesu} at page {p_index}")
                    else:
                        all_data.extend(items)
                        print(f"Collected {len(items)} items from {api['name']} DAESU {daesu} page {p_index}")
                        p_index += 1
                    
                    # Add delay to avoid overwhelming the server
                    await asyncio.sleep(0.1)
                    
            except Exception as error:
                print(f"Error fetching {api['name']} DAESU {daesu} page {p_index}: {error}")
                has_more_data = False

        return {
            'api': api['name'],
            'daesu': daesu,
            'total_items': len(all_data),
            'data': all_data
        }

    async def fetch_age_data(self, session: aiohttp.ClientSession, api: Dict[str, Any], 
                           age: int) -> Dict[str, Any]:
        """Fetch data for a specific AGE (assembly term)"""
        print(f"Starting data collection for {api['name']} AGE {age}...")
        
        all_data = []
        p_index = api.get('start_index', 1)
        has_more_data = True

        while has_more_data:
            try:
                url = f"{api['url']}?KEY={self.api_key}&Type=json&pIndex={p_index}&pSize={self.page_size}&AGE={age}"
                print(f"Fetching {api['name']} AGE {age} - Page {p_index}...")
                
                async with session.get(url) as response:
                    response.raise_for_status()
                    data = await response.json()
                    
                    # Extract the actual data array from the response
                    data_key = next(iter(data.keys()))
                    items = data.get(data_key, [None, {}])[1].get('row', []) if len(data.get(data_key, [])) > 1 else []
                    
                    if not items:
                        has_more_data = False
                        print(f"No more data for {api['name']} AGE {age} at page {p_index}")
                    else:
                        all_data.extend(items)
                        print(f"Collected {len(items)} items from {api['name']} AGE {age} page {p_index}")
                        p_index += 1
                    
                    # Add delay to avoid overwhelming the server
                    await asyncio.sleep(0.1)
                    
            except Exception as error:
                print(f"Error fetching {api['name']} AGE {age} page {p_index}: {error}")
                has_more_data = False

        return {
            'api': api['name'],
            'age': age,
            'total_items': len(all_data),
            'data': all_data
        }

    def save_to_file(self, result: Dict[str, Any], filename: str) -> None:
        """Save result data to JSON file"""
        try:
            file_path = self.base_dir / filename
            with open(file_path, 'w', encoding='utf-8') as f:
                json.dump(result, f, ensure_ascii=False, indent=2)
            print(f"Saved {result['total_items']} items to {filename}")
        except Exception as error:
            print(f"Error saving to {filename}: {error}")

    async def run(self) -> None:
        """Main execution method"""
        print('Starting API data aggregation...')
        
        connector = aiohttp.TCPConnector(limit=10)
        timeout = aiohttp.ClientTimeout(total=30)
        
        async with aiohttp.ClientSession(connector=connector, timeout=timeout) as session:
            for api in self.apis:
                try:
                    if api.get('is_daesu_iteration'):
                        # Handle DAESU iteration for member history API
                        for daesu in range(10, 23):  # 10-22대
                            result = await self.fetch_daesu_data(session, api, daesu)
                            filename = api['filename'].replace('{DAESU}', str(daesu))
                            self.save_to_file(result, filename)
                            print(f"Completed {api['name']} DAESU {daesu}: {result['total_items']} total items\n")
                    
                    elif api.get('is_age_iteration'):
                        # Handle AGE iteration for bills API
                        for age in range(10, 23):  # 10-22대
                            result = await self.fetch_age_data(session, api, age)
                            filename = api['filename'].replace('{AGE}', str(age))
                            self.save_to_file(result, filename)
                            print(f"Completed {api['name']} AGE {age}: {result['total_items']} total items\n")
                    
                    else:
                        result = await self.fetch_api_data(session, api)
                        self.save_to_file(result, api['filename'])
                        print(f"Completed {api['name']}: {result['total_items']} total items\n")
                        
                except Exception as error:
                    print(f"Failed to process {api['name']}: {error}")
        
        print('All APIs processed successfully!')

def main():
    """Main function to run the assembly data fetcher"""
    fetcher = AssemblyDataFetcher()
    asyncio.run(fetcher.run())

if __name__ == "__main__":
    main()